FirebaseAuth/Sources/Utilities/FIRAuthRecaptchaVerifier.m

```diff
@@ -0,0 +1,255 @@
+      @(FIRAuthRecaptchaActionSignUpPassword) : @"signUpPassword"
+    };
+  });
+  if (sharedRecaptchaVerifier.auth != auth) {
+    sharedRecaptchaVerifier.agentConfig = nil;
+    sharedRecaptchaVerifier.tenantConfigs = nil;
+    sharedRecaptchaVerifier.auth = auth;
+  }
+  return sharedRecaptchaVerifier;
+}
+
+- (NSString *)siteKey {
+  if ([FIRAuth auth].tenantID == nil) {
+    return self->_agentConfig.siteKey;
+  } else {
+    return self->_tenantConfigs[[FIRAuth auth].tenantID].siteKey;
+  }
+}
+
+#if TARGET_OS_IOS && !TARGET_OS_MACCATALYST
+- (BOOL)enablementStatusForProvider:(FIRAuthRecaptchaProvider)provider {
+  if ([FIRAuth auth].tenantID == nil) {
+    return [self->_agentConfig.enablementStatus[providerToStringMap[@(provider)]] boolValue];
+  } else {
+    return [self->_tenantConfigs[[FIRAuth auth].tenantID]
+                .enablementStatus[providerToStringMap[@(provider)]] boolValue];
+  }
+}
+
+- (void)verifyForceRefresh:(BOOL)forceRefresh
+                    action:(FIRAuthRecaptchaAction)action
+                completion:(nullable FIRAuthRecaptchaTokenCallback)completion {
+  [self
+      retrieveRecaptchaConfigForceRefresh:forceRefresh
+                               completion:^(NSError *_Nullable error) {
+                                 if (error) {
+                                   completion(nil, error);
+                                 }
+                                 if (!self.recaptchaClient) {
+                                   NSString *siteKey = @"";
+                                   if ([FIRAuth auth].tenantID == nil) {
+                                     siteKey = self->_agentConfig.siteKey;
+                                   } else {
+                                     siteKey =
+                                         self->_tenantConfigs[[FIRAuth auth].tenantID].siteKey;
+                                   }
+                                   Class RecaptchaClass = NSClassFromString(@"Recaptcha");
+                                   if (RecaptchaClass) {
+                                     SEL selectorWithTimeout = NSSelectorFromString(
+                                         @"getClientWithSiteKey:withTimeout:completion:");
+                                     if ([RecaptchaClass respondsToSelector:selectorWithTimeout]) {
+                                       void (*funcWithTimeout)(
+                                           id, SEL, NSString *, double,
+                                           void (^)(
+                                               id<RCARecaptchaClientProtocol> _Nullable recaptchaClient,
+                                               NSError *_Nullable error)) = (void *)[RecaptchaClass
+                                           methodForSelector:selectorWithTimeout];
+                                       funcWithTimeout(
+                                           RecaptchaClass, selectorWithTimeout, siteKey, 10000.0,
```

Comment:
Is this timeout in milliseconds? What is it for?

Reply:
> Is this timeout in milliseconds?
Yes.

The method would fail if the client creation takes more than 10 seconds (default value).